### PR TITLE
Replace WinRT clipboard API with Win32 for pasting

### LINKED ISSUE #14323: Paste from clipboard doesn't always work
### Windows Terminal version

1.15.2874.0

### Windows build number

10.0.19044.2130

### Other Software

Old Intel workbench internal tool
Notepad++ v8.4.1

### Steps to reproduce

Copy some text out of the old workbench tool and attempt to paste it into Terminal. Nothing is pasted. I *can* paste into Notepad++. If I do, I can then copy that text from Notepad++ and then paste it into terminal. Without knowing any details, it seems to me that Notepad++ is supporting some older clipboard methodology that is not present in Terminal. Please add whatever it is that is missing in Terminal.

### Expected Behavior

What was copied should have been pasted into Terminal as input.

### Actual Behavior

Nothing is pasted when data is copied from that old app.

### LINKED ISSUE #15315: Paste is annoyingly asynchronous wrt typeahead
### Windows Terminal version

1.17.1023

### Windows build number

10.0.19045.2846

### Other Software

git version 2.40.0.windows.1

### Steps to reproduce

Run Git Bash in Windows Terminal, while other processes are causing CPU and paging load.

Copy some text to the Clipboard from another application. (For example, copy a Git commit ID.)

In Git Bash, type some characters, press Shift+Insert to paste, and type more characters. (For example, type `git cherry-pick ` initially, and press Enter at the end.)

### Expected Behavior

Pasted text should appear in the correct place between the first and second sequence of characters.

### Actual Behavior

Pasted text occasionally appears after the second sequence of characters.  As if Shift+Insert starts an asynchronous paste operation that completes at some later time and injects its result into the terminal input at that time.

### PR DESCRIPTION
**Pull request by lhecker**
The Win32 API is significantly faster than the WinRT one, in the order
of around 300-1000x depending on the CPU and CPU load.

This might slightly improve the situation around #15315, but I suspect
that it requires many more fixes. For instance, we don't really have a
single text input "queue" into which we write. Multiple routines that
`resume_background` just to `WriteFile` into the input pipe are thus
racing against each other, contributing to the laggy feeling.
I also fear that the modern Windows text stack might be inherently
RPC based too, producing worse lag with rising CPU load.

This might fix #14323

## Validation Steps Performed
* Paste text from Edge ✅
* Paste text from Notepad ✅
* Right click the address bar in Explorer, choose "Copy address",
  paste text into WT ✅

### CODE REVIEW

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
-        const auto data = Clipboard::GetContent();
-
-        // This will switch the execution of the function to a background (not
-        // UI) thread. This is IMPORTANT, because the getting the clipboard data
-        // will crash on the UI thread, because the main thread is a STA.
-        co_await winrt::resume_background();
+        // Counterintuitively pasting the clipboard should block the UI thread, because that's
+        // actually the expectation of users: If you type something, paste and then type something again,
+        // text should be written in exactly that order, no matter how loaded the CPU is. See GH#15315.
+        assert(Dispatcher().HasThreadAccess());
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2535]: I would suggest suppressing whitespace changes. The indentation changed because I moved the try/catch up to surround the entire function.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 2]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
             }
 
-            constexpr const std::size_t minimumSizeForWarning = 1024 * 5; // 5 KiB
-            const auto warnLargeText = text.size() > minimumSizeForWarning &&
-                                       _settings.GlobalSettings().WarnAboutLargePaste();
+            const auto dataCleanup = wil::scope_exit([&]() {
+                GlobalUnlock(data);
+            });
+
+            text = winrt::hstring{ str, gsl::narrow_cast<uint32_t>(wcslen(str)) };
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2589]: I feel like the new code is very verbose, but also very easy to follow. If it's too verbose for you I wouldn't mind abstracting away the `OpenClipboard` / `CloseClipboard` calls into a safer wrapper like `wil::open_clipboard` (which lacks the retrying logic) and something like `wil::unique_hglobal_locked` (which unfortunately is `void*` and not `wchar_t*`).

IMO it's fine this way because this is the only use of this code and I wouldn't introduce an abstraction until we got 2-3 uses of this (at which point we'll probably have a better idea how the data types should look like, especially in respect to `SetClipboardData`).

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -2509,6 +2509,11 @@ namespace winrt::TerminalApp::implementation
         CATCH_LOG();
     }
 
+// The current UCRT has a neat AVX2 implementation for all of the various strlen() functions.
+// This greatly benefits _PasteFromClipboardHandler, where we're calling wcslen while holding a global
+// lock onto the clipboard. MSVC treats wcslen as a builtin and so we need to disable it explicitly.
+#pragma function(wcslen)
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2515]: _what,_ this is terrifying.

Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2515]: does this compile with clang?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
-        // UI) thread. This is IMPORTANT, because the getting the clipboard data
-        // will crash on the UI thread, because the main thread is a STA.
-        co_await winrt::resume_background();
+        // Counterintuitively pasting the clipboard should block the UI thread, because that's
+        // actually the expectation of users: If you type something, paste and then type something again,
+        // text should be written in exactly that order, no matter how loaded the CPU is. See GH#15315.
+        assert(Dispatcher().HasThreadAccess());
 
-        try
+        if (!IsClipboardFormatAvailable(CF_UNICODETEXT))
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2537]: what about `CF_OEMTEXT` and `CF_TEXT`?

I would guess that `StandardDataFormats::Text` handles that for us. Now we need to do it ourselves.

After fixing this, is it still as much faster?

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 3 above)
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2515]: Do you mean the clang-cl kind? As far as I can tell, clang never uses an intrinsic `wcslen` in the first place (which is nice, because why would it). For instance: https://godbolt.org/z/7rezdsWd9

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 4 above)
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2537]: > what about `CF_OEMTEXT` and `CF_TEXT`?

The 1992 API is actually way more 2023 than one might think. 😄 It has "synthesized types", so that if one application writes a `CF_OEMTEXT` you can still receive it as a `CV_UNICODE_TEXT` if that's what you're asking for. See here: https://learn.microsoft.com/en-us/windows/win32/dataxchg/clipboard-formats#synthesized-clipboard-formats

> After fixing this, is it still as much faster?

No fixing needed, but even if we did, it'll always be faster. The only "expensive" part of the old API is `OpenClipboard` which acquires an mutex in the kernel. That takes roughly 10us on my PC. All other calls are practically copying data straight out of it for free. The WinRT API on the other hand does 1 RPC call for anything you do on the `DataPackageView` object (3500us per call on my PC). So just the `.Contains(StandardDataFormats.Text)` alone is worth _thousands_ of win32 clipboard API calls.

> I would guess that `StandardDataFormats::Text` handles that for us. Now we need to do it ourselves.

Actually the "ClipboardServer" process is an entirely standalone thing that isn't built on top of the win32 API at all. Instead, it's basically... oh well... oh no...
Records indicate that ClipboardServer is actually just the _exact same thing as the win32 API_! It also runs as a single, global instance and it also is nothing but a `SRWLOCK` that protects a global, single agile instance of a `DataPackageView` that is then lent out to anyone calling `Clipboard::GetContent`. Software engineering is a circle. 😄

So yeah... it's actually just a mutex around a `DataPackageView`. I _think_ it stays in sync with the win32 side of things via the Windows Notification Facility (WNF), which, if true, would also be an absolute banger. 😄

Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2537]: @KalleOlaviNiemitalo Thanks! That's a good point. I think adding the `IsClipboardFormatAvailable` is a bit of a premature optimization anyways, I think. Usually when you right click into Windows Terminal you got something to paste in your clipboard already. 🤔

Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2537]: Actually, nevermind... It seems to work just fine. I've tested it using this application:
```cpp
#define NOMINMAX
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

int main() {
    OpenClipboard(nullptr);
    EmptyClipboard();

    const auto mem = GlobalAlloc(GMEM_MOVEABLE, 4);
    memcpy(GlobalLock(mem), "foo", 4);
    GlobalUnlock(mem);

    SetClipboardData(CF_TEXT, mem);
    CloseClipboard();
    return 0;
}
```

I can paste "foo" just fine in WT afterwards. (Also, I'm surprised that `OpenClipboard(nullptr)` works! That's pretty neat.)

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 5]
```diff
@@ -2509,6 +2509,11 @@ namespace winrt::TerminalApp::implementation
         CATCH_LOG();
     }
 
+// The current UCRT has a neat AVX2 implementation for all of the various strlen() functions.
+// This greatly benefits _PasteFromClipboardHandler, where we're calling wcslen while holding a global
+// lock onto the clipboard. MSVC treats wcslen as a builtin and so we need to disable it explicitly.
```
Comment (metathinker) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2514]: This is fine as is, but as a side note: If the clipboard data source app needs to delay render the clipboard data, it's less likely that a slow strlen() in the app reading the clipboard will account for a significant part of the time the clipboard lock is held.

**Review by metathinker — COMMENTED** [review]
Hi, former developer-maintainer of the Windows clipboard implementation here - sorry for the hit and run review, but when I see clipboard, I'm always a bit intrigued :)

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 6]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
-            else if (data.Contains(StandardDataFormats::StorageItems()))
+            co_return;
+        }
+
+        // The Win32 clipboard API allows us to synchronously retrieve the clipboard on the UI thread
+        // which is exactly what we want (see above). Additionally...
+        //
+        // The old Win32 clipboard API as used below is somewhere in the order of 1000x faster than the WinRT one,
+        // which has about the same latency as a fibre optics connection across the entire WA state.
+        // Don't use the WinRT clipboard API. RPC calls are not funny.
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2547]: I tested it using some code that I have already deleted. To make the new comparison as realistic as possible, I've chosen to implement the benchmark inside Windows Terminal itself and modified like so:
```cpp
fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/, const PasteFromClipboardEventArgs eventArgs)
try
{
    const auto beg = std::chrono::high_resolution_clock::now();
    winrt::hstring text;

    for (int i = 0; i < 1000; i++)
    {
        const auto data = Clipboard::GetContent();

        if (data.Contains(StandardDataFormats::Text()))
        {
            text = co_await data.GetTextAsync();
        }
        // Windows Explorer's "Copy address" menu item stores a StorageItem in the clipboard, and no text.
        else if (data.Contains(StandardDataFormats::StorageItems()))
        {
            auto items = co_await data.GetStorageItemsAsync();
            if (items.Size() > 0)
            {
                auto item = items.GetAt(0);
                text = item.Path();
            }
        }
    }

    const auto end = std::chrono::high_resolution_clock::now();
    const auto dur = std::chrono::duration<float, std::milli>(end - beg).count() / 1000.0f;
    wchar_t buffer[64];
    swprintf_s(buffer, L"%f\n", dur);
    OutputDebugStringW(&buffer[0]);

    // ... remaining contents of the function ...
}
CATCH_LOG();
```

Pasting text from X takes Y ms per iteration:
* Edge (Win32 clipboard API): 2.5ms
* Notepad: 3ms
* Explorer: 2.4ms
* The same instance of Windows Terminal (WinRT API): 2.1ms
* A different instance of Windows Terminal: also 2.1ms
* [My bodgy clipboard code](https://github.com/microsoft/terminal/pull/15360#discussion_r1194466002): 2.7ms

It seems there's no difference in latency for us between whether an application uses the WinRT or Win32 API for setting the clipboard contents. This made me curious whether the air got to my head once again and if I got everything, I said above wrong. So, I've tried to recreate my initial test code:
```cpp
#define NOMINMAX
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

// NOTE: This code relies on the latest cppwinrt version. To make matters simpler I've simply
// copy-pasted the winrt headers from terminal's "Generated Files" next to the current file.
#include "winrt/Windows.ApplicationModel.DataTransfer.h"
#include "winrt/Windows.Foundation.h"

int main() {
    winrt::init_apartment(winrt::apartment_type::single_threaded);

    []() -> winrt::fire_and_forget {
        const auto beg = std::chrono::high_resolution_clock::now();
        winrt::hstring text;

        for (int i = 0; i < 1000; i++) {
            const auto data = winrt::Windows::ApplicationModel::DataTransfer::Clipboard::GetContent();
            text = co_await data.GetTextAsync();
        }

        const auto end = std::chrono::high_resolution_clock::now();
        const auto dur = std::chrono::duration<float, std::milli>(end - beg).count() / 1000.0f;
        wprintf(L"%f\n", dur);
        PostQuitMessage(0);
    }();

    MSG msg;
    while (GetMessageW(&msg, nullptr, 0, 0)) {
        DispatchMessageW(&msg);
    }

    return 0;
}
```

Using it I get approximately the same latency as I get with Windows Terminal's code, which makes me believe that it's most likely correct, even if I can't immediately prove it. (If it's incorrect, please let me know!)

I did this because this allows me to run it in NVIDIA Nsight Systems to observe cross process behavior. Here's how it looks like when pasting from Edge into this application:
image

With that as an orientation, here's how one call (2.7ms) looks like:
image

You're right! There's no COM call in `GetTextAsync` when the clipboard content comes from Edge. `GetContent` however does involve 1 `CoCreateInstance` and 3 COM calls. I've blindly assumed that these are 1:1 directly related to the WinRT call that I'm doing. I'm genuinely sorry for this.

When the last instance of the `CDataPackageView` is destroyed it'll drop the `CNonAgileOleRefTaker`, which will end up calling `OleUninitialize`. This is probably not a good idea since this will also destroy the cached window it uses to call `OpenClipboard`. That's why every time you (or rather I, in my example project) call `Clipboard::GetContent` it'll recreate the window, which is very expensive overall (about half the entire CPU cost, the rest being mostly scheduling latency). Given that, I expect that the cost of the API would be significantly reduced if it wasn't calling `OleUninitialize`. And indeed, if I add a:
```cpp
if (OleInitialize(nullptr) != S_OK) {
    abort();
}
```
to the start of my test application, the cost of getting the clipboard contents from Edge is consistently reduced down to about 1.8ms. It is however still in stark contrast to the Win32 API which takes roughly 0.0066ms on average on my system. I didn't run a ETW trace where the remaining cost lies. But if there's any more interest in this topic, I'd be happy to trace it.

I do realize that it could be argued that the Win32 API has an unfair advantage since it doesn't need to do any async thread switching, the way the COM code needs to do it. As such I've made the following test based on this branch:
```cpp
fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/, const PasteFromClipboardEventArgs eventArgs)
try
{
    const auto beg = std::chrono::high_resolution_clock::now();
    winrt::hstring text;

    for (int i = 0; i < 1000; i++) // <--- added
    {
        co_await winrt::resume_background(); // <--- added

        // According to various reports on the internet, OpenClipboard might
        // fail to acquire the internal lock over RDP, due to rdpclip.exe.
        // I haven't seen this happen yet myself however.
        for (int attempts = 0;;)
        // ... the remaining code from this PR ...
        text = winrt::hstring{ str, gsl::narrow_cast<uint32_t>(wcslen(str)) };

        co_await winrt::resume_foreground(Dispatcher()); // <--- added
    }

    const auto end = std::chrono::high_resolution_clock::now();
    const auto dur = std::chrono::duration<float, std::milli>(end - beg).count() / 1000.0f;
    wchar_t buffer[64];
    swprintf_s(buffer, L"%f\n", dur);
    OutputDebugStringW(&buffer[0]);

    // ... remaining contents of the function from this PR ...
}
CATCH_LOG();
```

This increases the cost to roughly 0.1ms per iteration on my system.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 7]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
+        winrt::hstring text;
+        {
+            // According to various reports on the internet, OpenClipboard might
+            // fail to acquire the internal lock over RDP, due to rdpclip.exe.
+            // I haven't seen this happen yet myself however.
+            for (int attempts = 0;;)
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2553]: Good point, I'll change it to an linear backoff (cumulative exponential).

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 5 above)
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2514]: Yeah, I also benchmarked it, and for any realistic clipboard size (<100kB) it really doesn't matter at all. I'll remove it to keep the code base lean. But it does make a difference, even if it is small, due to the overall overhead of getting the clipboard content. It's about ~10x faster than the intrinsic `wcslen`.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 4 above)
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2537]: If you'd like, I think we should continue talking about these things in [the comment below](https://github.com/microsoft/terminal/pull/15360#discussion_r1194577685). Regarding `ClipboardServer.dll` however I'd like to already bring up that it is in fact involved when calling `Clipboard::GetContents`. Its actual runtime is so low though (compared to COM and OLE) that I couldn't figure out what exactly it's doing unfortunately. Here's the only traces I got for the brief ~3s test runtime:
```
Name                                                                                   Self, %    Stack, %    Module Name
Microsoft::WRL::FtmBase::MarshalInterface                                              ·          1.22        C:\Windows\System32\ClipboardServer.dll
Microsoft::WRL::Details::RuntimeClassImpl<...,IClipboardServer,...>::QueryInterface    0.07       0.09        C:\Windows\System32\ClipboardServer.dll
Microsoft::WRL::FtmBase::GetMarshalSizeMax                                             ·          0.04        C:\Windows\System32\ClipboardServer.dll
InlineIsEqualGUID                                                                      0.02       0.02        C:\Windows\System32\ClipboardServer.dll
ClipboardServer::GetBroker                                                             0.02       0.02        C:\Windows\System32\ClipboardServer.dll
```

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 8]
```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
...
+
+                break;
             }
 
-            if (_settings.GlobalSettings().TrimPaste())
+            const auto clipboardCleanup = wil::scope_exit([]() {
+                CloseClipboard();
+            });
+
+            const auto data = GetClipboardData(CF_UNICODETEXT);
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2573]: "Copy as path" does immediately set a `CF_UNICODETEXT`, as tested via [InsideClipboard](https://www.nirsoft.net/utils/inside_clipboard.html). But that comment is actually about a way less known feature in Explorer... Namely this context menu on the address bar:
image

The first item ("Copy address") sets the clipboard to contain these items:
* `CF_HDROP`
* `FileName`
* `FileNameW`
* `CF_LOCALE`

The clipboard doesn't contain a `CF_UNICODETEXT`, or anything similar at all. I do not understand how the translation to a `CF_UNICODETEXT` works in this case, but in my testing, I found that it works incredibly consistent. I suspect this might be because this Explorer functionality is incredibly old at this point (= was written around the time the win32 clipboard API was still maintained).

[the remark this answers is not in the document]
Comment (metathinker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2559]: My guess is the system (user32.dll or win32kfull.sys) can take CF_HDROP and synthesize CF_WhateverTEXT, at least when just one file is referenced in the CF_HDROP.

It's OK to close this thread out; it's at your discretion whether to add the comment I suggested in point 1.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (metathinker) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Those benchmarks seem complete enough, I'll accept them, and you can close this thread out. :)

I raised this point because I don't like to see people go back to the win32 clipboard API from UWP clipboard API. Win32 clipboard is really much harder to use correctly; even in a simple case like this you had to build a OpenClipboard busy wait loop which isn't even guaranteed to succeed.

But clearly the speed up is significant enough that it's worth making this change.

[the remark this answers is not in the document]
Comment (metathinker) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Re the other discussion about ClipboardServer.dll - I might be wrong or out of date to say that it doesn't get used for _basic_ clipboard functions. But my guess would be that its usage when you call UWP clipboard on desktop is for 1 of 2 reasons:

1. UWP app processes in a AppContainer use a broker to write to clipboard to filter out attempts to hijack higher-privileged processes with corrupt clipboard data. I thought that function wasn't built into ClipboardServer.dll, but maybe it is, or has become so. Within Windows Terminal, the UWP clipboard code in the caller process should be smart enough to see that it's not running from an AppContainer process and make the calls directly, but maybe it's not or there are other constraints.
2. On desktop, ClipboardServer is used to enforce clipboard enterprise data watermarking and restriction (Windows Information Protection/Enterprise Data Protection). This protection should apply to all clipboard APIs but I don't know how or when it's enforced for usage of each API.

[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: > I raised this point because I don't like to see people go back to the win32 clipboard API from UWP clipboard API. Win32 clipboard is really much harder to use correctly; even in a simple case like this you had to build a OpenClipboard busy wait loop which isn't even guaranteed to succeed.

Yeah, that's a totally fair point and I wouldn't mind sticking with the old code. In fact, the new API is honestly way nicer and easier to use. It's genuinely just not fast enough and I think this ever so slightly contributes to the growing computational inefficiency affecting Windows.

If there's an API that's a little bit harder to use, but multiple magnitudes more power efficient [^1], I think we not just should use it, but almost have a moral obligation to use it. Not just because we engineers are fickler when it comes to performance than product owners might hope [^2], but more importantly because the rising layers of abstraction and corresponding computational latency are unsustainable long-term. [^3]

In other words, I want Windows to be fast and power efficient, where everything you do is as good as it gets to the fullest extent of the underlying hardware and what we MS engineers are capable of delivering and it's what I'm trying to work towards.

[^1]: The code I've ended up with only needs to do a single `resume_background`. This ends up with a latency of just 13us per iteration on my system, or roughly 200x more efficient.
[^2]: Everyone immediately switched to Chrome when it came out, because it was much leaner than IE, same for VS Code and Atom. Why would this be any different for Windows and Proton/Wine one day?
[^3]: If the WinRT API had replaced the Win32 one, so that the latter is just a thin shim, I'd totally get it, but this way? When will we get rid of the original code?

**Comment by lhecker** [comment]
> I am curious about the "what if the source hangs while we're reading it" question - maybe there's a space for us to wrap this up as an AsyncAction that we can cancel if it doesn't return.

I recently changed this PR to yield to the background thread and untagged #15315 as being fixed, so it won't hang the UI anymore. We can't use an `AsyncAction`, because there's no way to cancel reading from the clipboard. It'll simply time out after 30s (as guaranteed by the win32 API). To us this doesn't really matter, because WinRT calls OLE and OLE calls the Win32 API which also hangs for 30s. It's a matryoshka.

> That's probably not as expensive as all the other COM calls.

Not even remotely. Even with yielding to a background thread first the total latency is still 20000% faster than the WinRT API on average (using percentages makes it look a bit funnier 😄).

> Honestly though I'm not super familiar with the edge cases with the win32 API so I'm not sure what we have to be afraid of.

The worst that can happen is that it works as bad as the conhost implementation, but the only major complaint we've ever gotten there is that pasting breaks Unicode, but that has always been our own fault (`TextToKeyEvents` being UCS-2). On that note I've just noticed that I should use `wcsnlen` and not `wcslen` to protect ourselves against strings without null terminator.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 9]
```diff
@@ -2520,62 +2520,91 @@ namespace winrt::TerminalApp::implementation
...
-                text = { Utils::TrimPaste(text) };
-                if (text.empty())
+                if (attempts > 5)
                 {
-                    // Text is all white space, nothing to paste
                     co_return;
                 }
+
+                attempts++;
+                Sleep(10 * attempts);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2552]: I get that we should wait before trying again, but why are we doing `10*attempts` here?

**Review by carlos-zamora — APPROVED** [review]

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 9 above)
Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2552]: The usual approach is to do exponential backoffs: Every iteration you increase the backoff duration by like 2x for instance or something. I felt like doing just a linear backoff would work well enough for our primitive use case. After all, conhost uses the same API and doesn't use any retrying _at all_.

**Comment by lhecker** [comment]
@zadjii-msft @DHowett It'd be great if we could still get this into 1.19. We've been testing it for quite a bit already after all. It's missing just 1 more approval.

**Review by zadjii-msft — APPROVED** [review]
sure let's give it a shot

**Comment by zadjii-msft** [comment]
This one is from before the CI rewrite so it might need to be manually merged...

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 10]
```diff
@@ -2640,62 +2640,91 @@ namespace winrt::TerminalApp::implementation
...
-                warnMultiLine = hasNewLine;
+                co_return;
             }
 
-            constexpr const std::size_t minimumSizeForWarning = 1024 * 5; // 5 KiB
-            const auto warnLargeText = text.size() > minimumSizeForWarning &&
-                                       _settings.GlobalSettings().WarnAboutLargePaste();
+            const auto dataCleanup = wil::scope_exit([&]() {
+                GlobalUnlock(data);
+            });
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2693]: technically there's `wil::unique_hglobal_locked`, which takes an HGLOBAL and unlocks it later! On its own!

**Review by DHowett — APPROVED** [review]
